pinecone/core/grpc/index_grpc.py

```diff
@@ -224,16 +225,19 @@ def _vector_transform(item):
 
         request = UpsertRequest(vectors=list(map(_vector_transform, vectors)), **kwargs)
         timeout = kwargs.pop('timeout', None)
-
-        response = self._wrap_grpc_call(self.stub.Upsert, request, timeout=timeout)
-        return parse_upsert_response(response)
+        if async_req:
+            return self._wrap_grpc_call(self.stub.Upsert.future, request, timeout=timeout)
+        else:
+            return self._wrap_grpc_call(self.stub.Upsert, request, timeout=timeout)
```

Comment:
I don't think the exception handling logic in _wrap_grpc_call will still apply correctly in the async case. Maybe we need to wrap the grpc future in a future of our own which delegates to the grpc one but does error handling when results are retrieved.

You could technically even add a "no 3p calls" integ test (creating a local VectorService grpc server as in https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests/unit/_metadata_test.py#L178-L189) to test timeout behaviors and other grpc client things, but only if you think it's worth the effort. Maybe we should just rely on our core suite of integ tests for that, at least for now.